src/app/clusters/zone-management-server/zone-management-server.cpp

```diff
@@ -0,0 +1,790 @@
+
+    if (DoesZoneAlreadyExist(zoneToUpdate.use, twoDCartVertices, DataModel::MakeNullable<uint16_t>(zoneID)))
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::AlreadyExists);
+        return;
+    }
+
+    if (ZoneUtil::IsZoneSelfIntersecting(twoDCartVertices))
+    {
+        ChipLogError(Zcl, "HandleUpdateTwoDCartesianZone: Found self-intersecting polygon vertices for zone");
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::DynamicConstraintError);
+        return;
+    }
+
+    TwoDCartesianZoneStorage twoDCartZoneStorage;
+    twoDCartZoneStorage.Set(zoneToUpdate.name, zoneToUpdate.use, twoDCartVertices, zoneToUpdate.color);
+
+    // Call the delegate
+    status = mDelegate.UpdateTwoDCartesianZone(zoneID, twoDCartZoneStorage);
+    if (status == Status::Success)
+    {
+        ZoneInformationStorage zoneInfo;
+        zoneInfo.Set(zoneID, ZoneTypeEnum::kTwoDCARTZone, ZoneSourceEnum::kUser, MakeOptional(twoDCartZoneStorage));
+
+        UpdateZone(zoneID, zoneInfo);
+    }
+    else
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        return;
+    }
+
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Success);
+}
+
+void ZoneMgmtServer::HandleRemoveZone(HandlerContext & ctx, const Commands::RemoveZone::DecodableType & commandData)
+{
+    uint16_t zoneID = commandData.zoneID;
+
+    auto foundZone =
+        std::find_if(mZones.begin(), mZones.end(), [&](const ZoneInformationStruct & z) { return z.zoneID == zoneID; });
+    if (foundZone == mZones.end())
+    {
+        ChipLogError(Zcl, "ZoneMgmt[ep=%d]: No zone exists by id %d", mEndpointId, zoneID);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::NotFound);
+        return;
+    }
+
+    if (foundZone->zoneSource == ZoneSourceEnum::kMfg)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+        return;
+    }
+
+    auto foundTrigger =
+        std::find_if(mTriggers.begin(), mTriggers.end(), [&](const ZoneTriggerControlStruct & t) { return t.zoneID == zoneID; });
+    if (foundTrigger != mTriggers.end())
+    {
+        ChipLogError(Zcl, "ZoneMgmt[ep=%d]: Zone id %d is referenced in zone triggers", mEndpointId, zoneID);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::InvalidInState);
+        return;
+    }
+
+    // Call the delegate
+    Status status = mDelegate.RemoveZone(zoneID);
+
+    if (status == Status::Success)
+    {
+        RemoveZone(zoneID);
+        mUserDefinedZonesCount--;
+    }
+    else
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        return;
+    }
+
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Success);
+}
+
+void ZoneMgmtServer::HandleCreateOrUpdateTrigger(HandlerContext & ctx,
+                                                 const Commands::CreateOrUpdateTrigger::DecodableType & commandData)
+{
+    ZoneTriggerControlStruct trigger = commandData.trigger;
```

Comment:
Where is the constraint validation on the fields of this struct?

Reply:
Missed adding them. Thanks. Fixed.